go-controller/pkg/network-controller-manager/netattachdefcontroller.go

```diff
@@ -0,0 +1,523 @@
+}
+
+func (nadController *netAttachDefinitionController) onNetworkAttachDefinitionUpdate(oldObj, newObj interface{}) {
+	oldNad := oldObj.(*nettypes.NetworkAttachmentDefinition)
+	newNad := newObj.(*nettypes.NetworkAttachmentDefinition)
+	if oldNad == nil || newNad == nil {
+		utilruntime.HandleError(fmt.Errorf("invalid net-attach-def provided to onNetworkAttachDefinitionUpdate()"))
+		return
+	}
+
+	// don't process resync or objects that are marked for deletion
+	if oldNad.ResourceVersion == newNad.ResourceVersion ||
+		!newNad.GetDeletionTimestamp().IsZero() {
+		return
+	}
+
+	klog.V(4).Infof("Updating net-attach-def %s/%s", newNad.Namespace, newNad.Name)
+	nadController.queueNetworkAttachDefinition(newObj)
+}
+
+func (nadController *netAttachDefinitionController) onNetworkAttachDefinitionDelete(obj interface{}) {
+	nad := obj.(*nettypes.NetworkAttachmentDefinition)
+	if nad == nil {
+		utilruntime.HandleError(fmt.Errorf("invalid net-attach-def provided to onNetworkAttachDefinitionDelete()"))
+		return
+	}
+
+	klog.V(4).Infof("Deleting net-attach-def %s/%s", nad.Namespace, nad.Name)
+	nadController.queueNetworkAttachDefinition(obj)
+}
+
+// GetAllNetworkControllers returns a snapshot of all managed nad associated network controllers.
+// Caller needs to note that there are no guarantees the return results reflect the real time
+// condition. There maybe more controllers being added, and returned controllers may be deleted
+func (nadController *netAttachDefinitionController) GetAllNetworkControllers() []NetworkController {
+	allNetworkNames := nadController.perNetworkNadNameInfo.GetKeys()
+	allNetworkControllers := make([]NetworkController, 0, len(allNetworkNames))
+	for _, netName := range allNetworkNames {
+		nadController.perNetworkNadNameInfo.LockKey(netName)
+		nni, ok := nadController.perNetworkNadNameInfo.Load(netName)
+		if ok {
+			allNetworkControllers = append(allNetworkControllers, nni.oc)
+		}
+		nadController.perNetworkNadNameInfo.UnlockKey(netName)
+	}
+	return allNetworkControllers
+}
+
+// AddNetAttachDef adds the given nad to the associated controller. It creates the controller if this
+// is the first nad of the network.
+// Non-retriable errors (configuration error etc.) are just logged, and the function immediately returns nil.
+func (nadController *netAttachDefinitionController) AddNetAttachDef(ncm NetworkControllerManager,
+	netattachdef *nettypes.NetworkAttachmentDefinition, doStart bool) error {
+	var netConfInfo util.NetConfInfo
+	var nInfo util.NetInfo
+	var err, invalidNadErr error
+	var netName string
+
+	netAttachDefName := util.GetNadName(netattachdef.Namespace, netattachdef.Name)
+	klog.Infof("Add net-attach-def %s", netAttachDefName)
+
+	nInfo, netConfInfo, invalidNadErr = util.ParseNADInfo(netattachdef)
+	if invalidNadErr == nil {
+		netName = nInfo.GetNetworkName()
+		if netName == types.DefaultNetworkName {
+			invalidNadErr = fmt.Errorf("NAD for default network, skip it")
```

Comment:
because this may be an update of a nad that already asssoicated with an controller. In that case, we'd like to delete the old nad from the controller first.

Reply:
we can say so, the result would be unexpected.